### ISSUE TITLE
Trying to upgrade/downgrade Dell Force10 Switch via netmiko. The reload output buffer fails the script

### ISSUE THREAD

**Issue by tamsel2017** [ktbyers/netmiko#646]
Hi Experts,

I am trying to upgrade or downgrade a Dell Force10 Switch with a input image name and then to reload the switch to verify if the switch has booted using the input image name via netmiko library using SSH.
Steps I follow:
1. configure the boot image
2. wr memory
3. reload 
         - in reload there is a interactive dialog as follows:
         System configuration has been modified. Save? [yes/no]: yes
         (or)
         Proceed with reload [confirm yes/no]: yes

**For reload command I use the following code, and it is successful:**
cmds_4 = 'do reload'
output_buffer_1 = remote_conn.send_command_timing(cmds, delay_factor=2)

**when the interactive dialog starts, I face the issue in the below code:**
output_buffer_3 = remote_conn.**send_command_expect**(cmds_3, expect_string='password:', max_loops=250, delay_factor=2)
**_(or)_**
output_buffer = remote_conn.**send_command_timing**(cmds_3, max_loops=250, delay_factor=2)

**Error which I am getting is as follows:**
> working_files/upgrade/device.py(63)execute()
-> output_buffer_3 = remote_conn.send_command_expect(cmds_3, expect_string='password:', max_loops=250, delay_factor=2)
(Pdb) n
IOError: IOError(...sword:',)
> /working_files/upgrade/device.py(63)execute()

I am expecting the Switch to reload and receive the new SSH prompt, where I want to provide the username/password and enter the "enable" mode and verify if the switch (show command -  "show boot system stack-unit all")has booted with the image name which is provided in the input variable.

Please let me know if this is possible via netmiko library and also your thoughts on the error,
======================================
Script Code:
cmds_4 = 'do reload'
output_buffer_1 = remote_conn.send_command_timing(cmds_4, delay_factor=2)
print output_buffer_1
sys_conf_compile = re.escape('System configuration has been modified. Save? [yes/no]:')
output_regex_1 = re.findall(sys_conf_compile, output_buffer_1)
proceed_compile = re.escape('Proceed with reload [confirm yes/no]:')
output_regex_2 = re.findall(proceed_compile, output_buffer_1)
if output_regex_1 != []:
    if str(output_regex_1[0]) == 'System configuration has been modified. Save? [yes/no]:':
            cmds_2 = 'yes'
             output_buffer_3 = remote_conn.send_command_expect(cmds_2, expect_string='password:', max_loops=250, delay_factor=2)

================================================
Automated Script Error during execution:
(Pdb) n
> /working_files/upgrade/device.py(63)execute()
-> output_buffer_3 = remote_conn.send_command_expect(cmds_3, expect_string='password:', max_loops=250, delay_factor=2)
(Pdb) n
IOError: IOError(...sword:',)
> /working_files/upgrade/device.py(63)execute()
-> output_buffer_3 = remote_conn.send_command_expect(cmds_3, expect_string='password:', max_loops=250, delay_factor=2)

=============================================================

Manual Steps:

1. Configure the System to boot using Partition A or B
"boot  system stack-unit 1 primary system: b:"
(or)
"boot  system stack-unit 1 primary system: a:"
2. "write memory"

3. reload
Either:
System configuration has been modified. Save? [yes/no]: yes
(or)
Proceed with reload [confirm yes/no]: yes

4. After reboot verify the boot image via the show command
"show boot system stack-unit all"

================================================
Command for upgrade / downgrade:
"boot  system stack-unit 1 primary system: b:"
================================================
Show Command for listing the images available in Partition A and B:
"show boot system stack-unit all"
================================================
Manual SSH connection Steps to connect to the Switch:
bash-3.2$ ssh admin@xx.xx.xx.xx
admin@xx.xx.xx.xx's password:

**Comment by tamsel2017** [comment]
Hi Experts,

I modified the below code for **"expect_string"** as **'' (null string)** and its providing the buffer and the switch reload process starts.

output_buffer_3 = remote_conn.send_command_expect(cmds_3, expect_string='password:', max_loops=250, delay_factor=2)

As a next step I am trying to initiate a new ssh connection via Netmiko and get the Switch prompt. As the switch is reloading  script fails with the following error when it tries to connect. Is there is any method to avoid the below error and put in  a loop for repeated try and establish the connection:
==================================================================
-> dut1_connect = system_connect()
(Pdb) n
NetMikoTimeoutException: NetMikoT...xxx:xx',)
> /upgrade/upgrade_downgrade_ssh_ftos.py(96)upgrade_downgrade_ssh_ftos()

**Comment by tamsel2017** [comment]
Hi Team,

Any thoughts on getting the switch prompt after reload operation ?

Thanks in Advance

**Comment by tamsel2017** [comment]
Hi Team,

The reload is successful. When the switch is booting up and when the script tries to establish ssh connection using netmiko, I am getting the below error

Traceback (most recent call last):
  File "upgrade_downgrade_ssh_ftos.py", line 7, in <module>
    dut1 = Dut('xx.xx.xx.xx', 'admin', 'admin', 'secretpass')
  File "/working_files/upgrade/dut.py", line 12, in __init__
    self.timeout = timeout
  File "/working_files/upgrade/dut.py", line 33, in connect
    self.conn = ConnectHandler(**dell_force10)
  File "build/bdist.linux-x86_64/egg/netmiko/ssh_dispatcher.py", line 131, in ConnectHandler
  File "build/bdist.linux-x86_64/egg/netmiko/base_connection.py", line 151, in __init__
  File "build/bdist.linux-x86_64/egg/netmiko/base_connection.py", line 536, in establish_connection
netmiko.ssh_exception.NetMikoTimeoutException: Connection to device timed-out: dell_force10 xx.xx.xx.xx:22

PLease let me know your thoughts.

**Comment by ktbyers** [comment]
@tamsel2017 The last error is because Netmiko is unable to connect to port 22 on the device. Have you verified the switch has fully rebooted and that you can manually SSH into (in other words, have you given enough time in your program for the switch reload to fully complete).

**Comment by tamsel2017** [comment]
Hi Kirk,

Thanks.

I have provide the time as follows in my code

dut1 = Dut('xx.xx.xx.xx', 'admin', 'admin', 'secretpass', timeout=300, session_timeout=300, keepalive=1, delay=180)

In the Class DUT, I have provided the following params
dell_force10 = {
                'device_type' : 'dell_force10',
                'ip' : self.ip,
                'username' : self.username,
                'password' : self.password,
                'secret' : self.secret,
                'verbose' : True,
                'timeout' : self.timeout,
                'keepalive' : self.keepalive,
                'session_timeout' : self.session_timeout,
                'use_keys' : False,
                'global_delay_factor' : self.delay,
                'port' : 22
                }
I am thinking of adding "time.sleep(60)" before the 2nd trigger of SSH connection.

Please let me know your thoughts.

**Comment by tamsel2017** [comment]
Hi KIrk,

I am trying a ping method approach for resolving the switch reload time.sleep(60).

Presently I am pinging the switch using subprocess once the switch reload takes place using a for loop and I compare if the ping is successful for a count of 60 to 70 and then I trigger a netmiko ssh connection. So far the subprocess call is working fine and I dont use time.sleep(60) as this is static and the switch reload will vary from model to model. 

I will post the subprocess call which I am using for ping in this forum. 

Thanks for your valid input.

**Comment by tamsel2017** [comment]
Hi Kirk,

I used the below Ping code in a for loop and set a count=70 and wait_time = 70 in the below code. The loop can be executed as required for any count.

ping_response = subprocess.Popen(["/bin/ping", "-c" + str(count), "-w" + str(count), ip_address_of_the_device], stdout=subprocess.PIPE).stdout.read()

I have used a if condition to stablize the ping using a consistent ping success of 70 or else to continue the loop

if int(ping_regexp[0]) == count:

This makes any of my devices stabilize irrespective of any delay after the reload.

Thanks
